interfaces/input_parameters/ring.py

```diff
@@ -182,182 +130,348 @@ class Ring:
         # based on the name
         if isinstance(Particle, beam.Particle):
             self.Particle = Particle
-        else:
+        elif isinstance(Particle, str):
             self.Particle = beam.make_particle(Particle)
-
-        # Reshaping the input synchronous data to the adequate format and
-        # get back the momentum program from RingOptions
-        if not isinstance(synchronous_data, dTypes._ring_program):
-            synchronous_data = \
-                dTypes._ring_program.conversions[func_type](synchronous_data)
-
-        if synchronous_data.shape[0] != len(self.ring_length):
-            raise excpt.InputDataError("ERROR in Ring: Number of sections " +
-                                       "and ring length size do not match!")
-
+        else:
+            raise excpt.InputError("The input Particle should be a " +
+                                   "Particle object or a str.")
+
+        # Getting all sections and checking their types
+        if isinstance(Section_list, RingSection):
+            Section_list = [Section_list]
+        if not isinstance(Section_list, list):
+            raise excpt.InputError("The Section_list should be a list or " +
+                                   "a single RingSection object instance.")
+        else:
+            for section in Section_list:
+                if not isinstance(section, RingSection):
+                    raise excpt.InputError(
+                        "The Section_list should be exclusively composed " +
+                        "of RingSection object instances.")
+
+                # Setting the Ring as attribute of the sections for reference
+                section._Ring = self
+
+        self.Section_list = Section_list
+        self.n_sections = len(self.Section_list)
+
+        # Extracting the length of sections to get circumference
+        self.section_length_design = np.array(
+            [section.length_design for section in self.Section_list])
+        self.circumference = np.sum(self.section_length_design)
+
+        # Computing ring radius
+        self.radius = self.circumference / (2 * np.pi)
+
+        # Extracting the bending radius from all sections
+        self.bending_radius = np.array([
+            section.bending_radius for section in self.Section_list])
+
+        # Extracting the synchronous data from the sections
+        for index_section, section in enumerate(self.Section_list):
+            if index_section == 0:
+                self.synchronous_data = section.synchronous_data
+            else:
+                if (section.synchronous_data == self.synchronous_data).all:
+                    pass
+                else:
+                    warn_message = 'The synchronous data for all sections ' + \
+                        'is not identical. This case is not yet fully ' + \
+                        'tested and is under implementation.'
+                    warnings.warn(warn_message)
```

Comment:
Shouldn't `self.synchronous_data` be an array at this point?  With `index_section` being used to populate it?

Reply:
This was modified in the more recent version, to be rechecked then

(the idea here was to explicitly raise exceptions for the cases not handled by the datatype.combine_function, better compromise could certainly be found)

Reply:
Nothing jumped out at me as being questionable about this in the new one, but I'll try and break it later and see what happens ;)